packages/orchestrator-ui-components/src/components/WFOWorkflowSteps/WFOWorkflowStepList/WFOWorkflowStepList.tsx

```diff
@@ -20,36 +20,53 @@ export const WFOWorkflowStepList = React.forwardRef(
 
         const t = useTranslations('processes.steps');
 
-        const allDetailsClosedState = new Map(
-            steps.map((_, index) => [index, false]),
-        );
-        const allDetailsOpenedState = new Map(
-            steps.map((_, index) => [index, true]),
+        const initialStepListItems: StepListItem[] = steps.map((step) => ({
+            step,
+            isExpanded: false,
+        }));
+        const [stepListItems, setStepListItems] =
+            useState(initialStepListItems);
+
+        const allStepsAreExpanded = stepListItems.every(
+            (item) => item.isExpanded,
         );
 
-        const [stepDetailStates, setStepDetailStates] = useState<
-            Map<number, boolean>
-        >(new Map(allDetailsClosedState));
+        const setExpandedStateStepListItems = (isExpanded: boolean) => {
+            setStepListItems(
+                stepListItems.map((item) => ({
+                    ...item,
+                    isExpanded,
+                })),
+            );
+        };
 
-        const toggleAllDetailsIsOpen = () => {
-            if (Array.from(stepDetailStates).every((s) => s[1])) {
-                setStepDetailStates(new Map(allDetailsClosedState));
-            } else {
-                setStepDetailStates(new Map(allDetailsOpenedState));
-            }
+        const handleToggleExpandedStepListItem = (
+            stepListItem: StepListItem,
+        ) => {
+            setStepListItems(
+                stepListItems.map((item) =>
+                    item.step === stepListItem.step
+                        ? {
+                              ...item,
+                              isExpanded: !item.isExpanded,
+                          }
+                        : item,
+                ),
+            );
         };
 
-        const toggleStepDetailIsOpen = (index: number) => {
-            if (stepDetailStates.has(index)) {
-                setStepDetailStates(
-                    new Map(
-                        stepDetailStates.set(
-                            index,
-                            !stepDetailStates.get(index),
-                        ),
-                    ),
-                );
-            }
+        const handleExpandStepListItem = (stepListItem: StepListItem) => {
+            setStepListItems(
+                stepListItems.map((item) => {
+                    if (item.step === stepListItem.step) {
```

Comment:
This looks quite similar to line 48-53

Reply:
extracted a function "updateStepListItem" to get rid of this code duplication